chain/network/src/snapshot_hosts/mod.rs

```diff
@@ -32,11 +36,143 @@ pub struct Config {
     /// least-recently-used cache. In the future, we may wish to
     /// implement something more sophisticated.
     pub snapshot_hosts_cache_size: u32,
+    /// The number of hosts we'll add to structures related to state part peer
+    /// selection each time we need to request parts from a new peer
+    pub part_selection_cache_batch_size: u32,
+}
+
+pub(crate) fn priority_score(peer_id: &PeerId, part_id: &PartId) -> [u8; 32] {
+    let mut h = Sha256::new();
+    h.update(peer_id.public_key().key_data());
+    h.update(part_id.idx.to_le_bytes());
+    h.finalize().into()
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct PartPriority {
+    score: [u8; 32],
+    // TODO: consider storing this on disk, so we can remember who hasn't
+    // been able to provide us with the parts across restarts
+    times_returned: usize,
+}
+
+impl PartPriority {
+    fn inc(&mut self) {
+        self.times_returned += 1;
+    }
+}
+
+impl PartialOrd for PartPriority {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PartPriority {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.times_returned.cmp(&other.times_returned).then_with(|| self.score.cmp(&other.score))
+    }
+}
+
+#[derive(Default)]
+struct PartPeerSelector {
+    peers: BTreeMap<PartPriority, PeerId>,
+}
+
+impl PartPeerSelector {
+    fn next(&mut self) -> Option<PeerId> {
+        match self.peers.pop_first() {
+            Some((mut priority, peer_id)) => {
+                priority.inc();
+                self.peers.insert(priority, peer_id.clone());
+                Some(peer_id)
+            }
+            None => None,
+        }
+    }
+
+    fn insert_peer(&mut self, peer_id: PeerId, score: [u8; 32]) {
+        self.peers.insert(PartPriority { score, times_returned: 0 }, peer_id);
+    }
+
+    fn len(&self) -> usize {
+        self.peers.len()
+    }
+
+    fn tried_everybody(&self) -> bool {
+        self.peers.iter().all(|(priority, _peer_id)| priority.times_returned > 0)
+    }
+}
+
+#[derive(Default)]
+struct PeerSelector {
+    selectors: HashMap<u64, PartPeerSelector>,
+}
+
+fn do_insert(
+    peers: &mut BTreeMap<[u8; 32], PeerId>,
+    peer_id: PeerId,
+    part_id: &PartId,
+    score: [u8; 32],
+) {
+    if let Some(other) = peers.insert(score, peer_id.clone()) {
+        // could handle it but it's so unlikely it's not really worth it. So this log statement serves
```

Comment:
I think if you change the representation to `BTreeMap<PartPriority, ()>`,  with `PartPriority` containing the `peer_id` you should be able to eliminate this problem.

Reply:
ohh good point